README.md

# `AnanseSeurat` package <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/JGASmits/AnanseSeurat/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/JGASmits/AnanseSeurat/actions/workflows/R-CMD-check.yaml)
[![codecov](https://codecov.io/github/JGASmits/AnanseSeurat/branch/main/graph/badge.svg?token=0XKWAD0KJ7)](https://app.codecov.io/github/JGASmits/AnanseSeurat)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/AnanseSeurat)](https://cran.r-project.org/package=AnanseSeurat)
<!-- badges: end -->

The `AnanseSeurat` package takes pre-processed clustered single cell
objects of scRNAseq and scATACseq or a multiome combination, and
generates files for gene regulatory network (GRN) analysis. It is part 
of the scANANSE pipeline. <https://doi.org/10.12688/f1000research.130530.1>

## Installation

`AnanseSeurat` can be installed from CRAN using

``` r
install.packages('AnanseSeurat')
```
Or to install the developmental branch from github:
``` r
library(devtools) # Tools to Make Developing R Packages Easier # Tools to Make Developing R Packages Easier
Sys.unsetenv("GITHUB_PAT")
remotes::install_github("JGASmits/AnanseSeurat@main")
```

### Usage

``` r
library("AnanseSeurat")
rds_file <- './scANANSE/preprocessed_PDMC.Rds'
pbmc <- readRDS(rds_file)
```

Next you can output the data from your single cell object, the file
format, config file and sample file are all ready to automate GRN
analysis using `anansnake`.
<https://github.com/vanheeringen-lab/anansnake>

``` r
export_CPM_scANANSE(
  pbmc,
  min_cells = 25,
  output_dir = './scANANSE/analysis',
  cluster_id = 'predicted.id',
  RNA_count_assay = 'RNA'
)

export_ATAC_scANANSE(
  pbmc,
  min_cells = 25,
  output_dir = './scANANSE/analysis',
  cluster_id = 'predicted.id',
  ATAC_peak_assay = 'peaks'
)

# Specify additional contrasts:
contrasts <-  c('B-naive_B-memory',
                'B-memory_B-naive',
                'B-naive_CD14-Mono',
                'CD14-Mono_B-naive')

config_scANANSE(
  pbmc,
  min_cells = 25,
  output_dir = './scANANSE/analysis',
  cluster_id = 'predicted.id',
  additional_contrasts = contrasts
)

DEGS_scANANSE(
  pbmc,
  min_cells = 25,
  output_dir = './scANANSE/analysis',
  cluster_id = 'predicted.id',
  additional_contrasts = contrasts
)
```

### install and run anansnake

Follow the instructions its respective github page,
<https://github.com/vanheeringen-lab/anansnake> After activating the
conda environment, use the generated files to run GRN analysis using
your single cell cluster data:

``` bash
anansnake \
--configfile scANANSE/analysis/config.yaml \
--resources mem_mb=48_000 --cores 12
```

### import ANANSE results back to your single cell object

After running Anansnake, you can import the TF influence scores back
into your single cell object of choice

``` r
pbmc <- import_seurat_scANANSE(pbmc,
                               cluster_id = 'predicted.id',
                               anansnake_inf_dir = "./scANANSE/analysis/influence")
TF_influence <- per_cluster_df(pbmc,
                               cluster_id = 'predicted.id',
                               assay = 'influence')
```

## Citation

> scANANSE gene regulatory network and motif analysis of single-cell clusters [version 1; peer review: awaiting peer review]
> Jos G.A. Smits, Julian A. Arts, Siebren Frölich, Rebecca R. Snabel, Branco M.H. Heuts, Joost H.A. Martens, Simon J van Heeringen, Huiqing Zhou
> F1000Research 2023, 12:243 (https://doi.org/10.12688/f1000research.130530.1) 

### Thanks to:

- Julian A. Arts and his Pycharm equivalent of this package
  <https://github.com/Arts-of-coding/AnanseScanpy>
- Siebren Frohlich and his anansnake implementation
  <https://github.com/vanheeringen-lab/anansnake>
- Rebecca R. Snabel for her implementation of the motif expression
  correlation analysis
- Branco Heuts for testing

### Credits

The hex sticker is generated using the
[`hexSticker`](https://github.com/GuangchuangYu/hexSticker) package.


NEWS.md
# AnanseSeurat 1.2.0
* Added Seurat V5 support
* Added motif array fix & matrix dependency fix
* Fixed config_scANANSE accepting both lists and vectors instead of only lists

# AnanseSeurat 1.1.0

* Added a `NEWS.md` file to track changes to the package.
* updated code formatting to adhere to the bestpractices() package
* updated Readme file
* added a smaller test file for the example of DEG_function to test within the allowed timeslot
* changed print statements to message(), warning() or stop() statements
* removed 'outputdir' variable from Maelstrom_Motif2TF, since it was not used and data is
  now returned to the S4 objects as assays (and no longer as output files)
* changed output_dir in all the export functions to be a mandatory argument instead of taking 
  the working dir as a default.
* added apache license to include patent info



tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(AnanseSeurat)

test_check("AnanseSeurat")


tests/testthat/test-export_functions.R
#test_functions

test_that("CPM & Count files are sucessfully generated", {
  outdir <- paste0(tempdir(),'/CPM')
  TPMfile <- paste0(outdir, '/TPM.tsv')
  Countfile <- paste0(outdir, '/RNA_Counts.tsv')
  on.exit(unlink(c(Countfile, TPMfile, outdir)))
  
  expect_false(file.exists(TPMfile))
  expect_false(file.exists(Countfile))
  
  #load dummy single cell object
  sce <- readRDS(testthat::test_path("sce_obj.Rds"))
  
  export_CPM_scANANSE(
    sce,
    min_cells <- 1,
    output_dir = outdir,
    cluster_id = 'seurat_clusters',
    RNA_count_assay = 'RNA'
  )
  expect_true(file.exists(TPMfile))
  expect_true(file.exists(Countfile))
  
  #check if the files are equal to the expected output
  TPM_correct <- read.table(testthat::test_path("TPM.tsv"))
  COUNT_correct <- read.table(testthat::test_path("RNA_Counts.tsv"))
  
  TPM_df <- read.table(TPMfile)
  COUNT_df <- read.table(Countfile)
  
  expect_true(all.equal(TPM_df, TPM_correct))
  expect_true(all.equal(COUNT_df, COUNT_correct))
})

test_that("Peak count matrix file is sucessfully generated", {
  outdir <- paste0(tempdir(),'/peaks')
  Peakfile <- paste0(outdir, '/Peak_Counts.tsv')
  on.exit(unlink(c(Peakfile, outdir)))
  
  expect_false(file.exists(Peakfile))
  
  #load dummy single cell object
  sce <- readRDS(testthat::test_path("sce_obj.Rds"))
  
  export_ATAC_scANANSE(
    sce,
    min_cells <- 1,
    output_dir = outdir,
    cluster_id = 'seurat_clusters',
    ATAC_peak_assay = 'peaks'
  )
  
  expect_true(file.exists(Peakfile))
  
  #check if the files are equal to the expected output
  Peak_correct <- read.table(testthat::test_path("Peak_Counts.tsv"))
  Peak_df <- read.table(Peakfile)
  expect_true(all.equal(Peak_df, Peak_correct))
})

test_that("Config and sample file are sucessfully generated", {
  outdir <- paste0(tempdir(),'/config')
  Configfile <- paste0(outdir, '/config.yaml')
  samplefile <- paste0(outdir, '/samplefile.tsv')
  
  on.exit(unlink(c(Configfile, samplefile, outdir)))
  
  expect_false(file.exists(Configfile))
  expect_false(file.exists(samplefile))
  
  
  #load dummy single cell object
  sce <- readRDS(testthat::test_path("sce_obj.Rds"))
  
  contrasts <- list('cluster1_cluster2',
                   'cluster2_cluster1',
                   'cluster3_cluster2')
  config_scANANSE(
    sce,
    min_cells <- 1,
    output_dir = outdir,
    cluster_id = 'seurat_clusters',
    genome = paste0(outdir, '/hg38'),
    additional_contrasts = contrasts
  )
  
  expect_true(file.exists(Configfile))
  expect_true(file.exists(samplefile))
  
  #check if the sample fileis  equal to the expected output
  samples_correct <-
    read.table(testthat::test_path("samplefile.tsv"))
  samples_df <- read.table(samplefile)
  expect_true(all.equal(samples_correct, samples_df))
})

test_that("Maelstrom count matrix is sucessfully generated", {
  outdir <- paste0(tempdir(),'/maelstrom')
  peakfile <- paste0(outdir, '/Peaks_scaled.tsv')
  on.exit(unlink(c(peakfile,outdir)))
  expect_false(file.exists(peakfile))
  
  #load dummy single cell object
  sce <- readRDS(testthat::test_path("sce_obj.Rds"))
  
  export_ATAC_maelstrom(
    sce,
    min_cells <- 1,
    output_dir = outdir,
    cluster_id = 'seurat_clusters',
    ATAC_peak_assay = 'peaks'
  )
  
  expect_true(file.exists(peakfile))
  
  #check if the sample fileis  equal to the expected output
  peaks_correct <-
    read.table(testthat::test_path("Peaks_scaled.tsv"))
  peaks_df <- read.table(peakfile)
  expect_true(all.equal(peaks_correct, peaks_df))
})


tests/testthat/test-import_functions.R
#test_functions

test_that("import seurat succesfully imports an influence file, wich contains an influence assay",
          {
            #load dummy single cell object
            sce <- readRDS(testthat::test_path("sce_obj.Rds"))
            sce <- import_seurat_scANANSE(
              sce,
              cluster_id = 'seurat_clusters',
              anansnake_inf_dir = testthat::test_path('influence/')
            )
            expect_true(length(sce@assays) == 3)
            expect_true('influence' %in% names(sce@assays))
            inf_df <- as.data.frame(sce@assays$influence@data)
            inf_correct <-
              read.table(testthat::test_path("influence_assay.tsv"))
            expect_true(all.equal(inf_df, inf_correct, tolerance = 1.0e-7))
            
          })

test_that(
  "import maelstrom succesfully imports a maelstrom file, wich contains an motif enrichment scores",
  {
    #load dummy single cell object
    sce <- readRDS(testthat::test_path("sce_obj.Rds"))
    sce <- import_seurat_maelstrom(
      sce,
      cluster_id = 'seurat_clusters',
      maelstrom_file = testthat::test_path('final.out.txt')
    )
    expect_true(length(sce@assays) == 3)
    expect_true('maelstrom' %in% names(sce@assays))
    maelstrom_df <- as.data.frame(sce@assays$maelstrom@data)
    maelstrom_correct <-
      read.table(testthat::test_path("maelstrom_table.tsv"))
    expect_true(all.equal(maelstrom_df, maelstrom_correct))
  }
)

test_that("per cell df works", {
  #load dummy single cell object
  sce <- readRDS(testthat::test_path("sce_inf_obj.Rds"))
  TF_influence <-
    per_cluster_df(sce, assay = 'influence', cluster_id = 'seurat_clusters')
  expect_true(all.equal(dim(TF_influence), c(4, 3)))
  
  TF_influence_correct <-
    read.table(testthat::test_path("influence_table.tsv"))
  expect_true(all.equal(TF_influence, TF_influence_correct))
})


R/calculate_DEGS.R
#' DEGS_scANANSE
#'
#' Calculate the differential genes needed for ananse influence
#' @param seurat_object seurat object
#' @param min_cells minimum of cells a cluster needs to be exported
#' @param output_dir directory where the files are outputted
#' @param cluster_id ID used for finding clusters of cells
#' @param RNA_count_assay assay containing the RNA data
#' @param additional_contrasts additional contrasts to add between clusters within cluster_ID
#' @export
DEGS_scANANSE <- function(seurat_object,
                          min_cells = 50,
                          output_dir = '~/',
                          cluster_id = 'seurat_clusters',
                          RNA_count_assay = "RNA",
                          additional_contrasts = 'None'
) {
  dir.create(file.path(paste0(output_dir,'/deseq2/')),showWarnings = FALSE)
  Seurat::Idents(seurat_object) <- cluster_id
  cluster_names = list()
  i = 1
  for (cluster in levels(Seurat::Idents(seurat_object))){
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells){
      cluster_names[i] <- cluster
      i = i+1 # Increase i to add clusters iteratively to the cluster_names list
    }}

  #lets generate the snakemake config file
  contrast_list <- as.list(paste0('anansesnake_',cluster_names,'_average'))
  if (typeof(additional_contrasts) == 'list'){print('adding additional contrasts')
    additional_contrasts <- paste0('anansesnake_',additional_contrasts)
    contrast_list = c(contrast_list,additional_contrasts)}

  for (contr in contrast_list){
    print(paste0('calculating DEGS for contrast ',contr))
    comparison1 <- stringr::str_split(contr, "_")[[1]][2]
    comparison2 <- stringr::str_split(contr, "_")[[1]][3]

    DEG_file <- paste0(output_dir,'/deseq2/','hg38-anansesnake_',comparison1,'_', comparison2,'.diffexp.tsv')

    if (file.exists(DEG_file)){print('skip')} else {

      if (comparison2 == 'average'){
        DEGS <- Seurat::FindMarkers(seurat_object,
                                    ident.1 = comparison1,
                                    assay = RNA_count_assay,
                                    logfc.threshold = 0.1,
                                    min.pct = 0.05,
                                    return.thresh = 0.1)
      }else{DEGS <- Seurat::FindMarkers(seurat_object,
                                        ident.1 = comparison1,
                                        ident.2 = comparison2,
                                        assay = RNA_count_assay,
                                        logfc.threshold = 0.1,
                                        min.pct = 0.05,
                                        return.thresh = 0.1)}
      DEGS <- DEGS[c('avg_log2FC','p_val_adj')]
      colnames(DEGS) <- c('log2FoldChange','padj')
      utils::write.table(DEGS,DEG_file, sep = '\t', quote = F)
    }
  }
}



R/calculateDEGS_functions.R
#' DEGS_scANANSE
#'
#' Calculate the differential genes needed for ananse influence
#' @param seurat_object seurat object
#' @param min_cells minimum of cells a cluster needs to be exported
#' @param output_dir directory where the files are outputted
#' @param cluster_id ID used for finding clusters of cells
#' @param genome path to the genome folder used for the anansnake config file
#' @param RNA_count_assay assay containing the RNA data
#' @param additional_contrasts additional contrasts to add
#' between clusters within cluster_ID
#' @return None, outputs DEG files in the output directory
#' @examples
#' sce_small <- readRDS(system.file("extdata","sce_obj_tiny.Rds",package = 'AnanseSeurat'))
#' DEGS_scANANSE(sce_small, min_cells = 2, output_dir = tempdir())
#' @export
DEGS_scANANSE <- function(seurat_object,
                          output_dir,
                          min_cells = 50,
                          cluster_id = 'seurat_clusters',
                          genome = './scANANSE/data/hg38',
                          RNA_count_assay = "RNA",
                          additional_contrasts = 'None') {
  if (missing(output_dir)) {
    warning('no output_dir specified')
  }
  dir.create(file.path(paste0(output_dir, '/deseq2/')), showWarnings = FALSE)
  Seurat::Idents(seurat_object) <- cluster_id
  cluster_names <- list()
  i <- 1
  for (cluster in levels(Seurat::Idents(seurat_object))) {
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells) {
      cluster_names[i] <- cluster
      i <- i + 1 # Increase i to add clusters iteratively
    }
  }
  
  #lets generate the snakemake config file
  contrast_list <-
    as.list(paste0('anansesnake_', cluster_names, '_average'))
  if (typeof(additional_contrasts) == 'list') {
    message('adding additional contrasts')
    additional_contrasts <-
      paste0('anansesnake_', additional_contrasts)
    contrast_list <- c(contrast_list, additional_contrasts)
  }
  
  for (contr in contrast_list) {
    message(paste0('calculating DEGS for contrast ', contr))
    comparison1 <- stringr::str_split(contr, "_")[[1]][2]
    comparison2 <- stringr::str_split(contr, "_")[[1]][3]
    
    DEG_file <-
      paste0(
        output_dir,
        '/deseq2/',
		basename(genome),
        '-anansesnake_',
        comparison1,
        '_',
        comparison2,
        '.diffexp.tsv'
      )
    
    if (file.exists(DEG_file)) {
      message('DEGfile already exist, skipping regenerating the DEG file')
      message('If new files are needed remove the existing DEG files ')
    } else {
      if (comparison2 == 'average') {
        DEGS <- Seurat::FindMarkers(
          seurat_object,
          ident.1 = comparison1,
          assay = RNA_count_assay,
          logfc.threshold = 0.1,
          min.pct = 0.05,
          return.thresh = 0.1
        )
      } else{
        DEGS <- Seurat::FindMarkers(
          seurat_object,
          ident.1 = comparison1,
          ident.2 = comparison2,
          assay = RNA_count_assay,
          logfc.threshold = 0.1,
          min.pct = 0.05,
          return.thresh = 0.1
        )
      }
      DEGS <- DEGS[c('avg_log2FC', 'p_val_adj')]
      colnames(DEGS) <- c('log2FoldChange', 'padj')
      utils::write.table(DEGS, DEG_file, sep = '\t', quote = FALSE)
    }
  }
}


R/ANANSE_seurat_wrapper_V2.R
#' export_CPM_scANANSE
#'
#' This functions exports CPM values from a seurat object
#' @param seurat object
#' @param min_cells: minimum of cells a cluster needs to be exported
#' @param output_dir: directory where the files are outputted
#' @param RNA_count_assay: assay of the seurat object containing the RNA count data
#' @param cluster_id: ID used for finding clusters of cells
#' @export
export_CPM_scANANSE <- function(seurat_object,
                                   min_cells = 50,
                                   output_dir = '~/',
                                   RNA_count_assay = "RNA",
                                   cluster_id = 'seurat_clusters'
) {
  dir.create(file.path(output_dir))
  Idents(seurat_object) <- cluster_id
  print('calculating CPM')
  seurat_object <- NormalizeData(seurat_object, 
                                 assay  = RNA_count_assay, 
                                 normalization.method = 'RC', 
                                 scale.factor = 1e6
  )
  rna_count_lists = list()
  FPKM_count_lists = list()
  cluster_names = list()
  i = 1
  for (cluster in levels(Idents(seurat_object))){
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells){
      print(paste0('gather data from ', cluster, ' with ', n_cells, ' cells'))
      cluster_names[i] <- cluster
      
      #lets grab the scRNAseq counts
      mat_counts <- seurat_object_cluster@assays[[RNA_count_assay]]@counts
      mat_counts <- as.matrix(rowSums(as.matrix(mat_counts)))
      colnames(mat_counts) <- cluster
      rna_count_lists[i] <- as.data.frame(mat_counts)
      
      #lets also grab the normalized intensities, wich we will
      #use as a FPKM like matrix, since the reads are UMI normalized
      #we do not expect a gene-length bias
      mat_norm <- seurat_object_cluster@assays[[RNA_count_assay]]@data
      mat_norm <- as.matrix(rowSums(as.matrix(mat_norm)))
      colnames(mat_norm) <- cluster
      FPKM_count_lists[i] <- as.data.frame(mat_norm)
      i = i + 1
    }else{print(paste0(cluster,' less than ',min_cells,' skipping'))}
  }
  
  #RNA count matrix
  count_matrix <- do.call(rbind, rna_count_lists)
  count_matrix <- as.data.frame(count_matrix)
  count_matrix <- t(count_matrix)
  colnames(count_matrix) <- cluster_names
  rownames(count_matrix) <- rownames(mat_counts)
  count_matrix <- as.data.frame(count_matrix)
  count_matrix$average <- round(rowMeans(count_matrix))
  
  #FPKM matrix
  FPKM_matrix <- do.call(rbind, FPKM_count_lists)
  FPKM_matrix <- as.data.frame(FPKM_matrix)
  FPKM_matrix <- t(FPKM_matrix)
  colnames(FPKM_matrix) <- cluster_names
  rownames(FPKM_matrix) <- rownames(mat_norm)
  FPKM_matrix <- as.data.frame(FPKM_matrix)
  FPKM_matrix$average <- rowMeans(FPKM_matrix)
  
  count_file <- paste(output_dir, "RNA_Counts.tsv", sep = '/')
  CPM_file <-paste(output_dir, "TPM.tsv", sep = '/')
  
  write.table(count_matrix, count_file, sep = '\t', quote = F)
  write.table(FPKM_matrix, CPM_file, sep = '\t', quote = F)
  return('done exporting cluster data files')
}
#' export_ATAC_scANANSE
#'
#' This functions exports ATAC values from a seurat object
#' @param seurat object
#' @param min_cells: minimum of cells a cluster needs to be exported
#' @param output_dir: directory where the files are outputted
#' @param ATAC_peak_assay: assay of the seurat object containing the peaks and peakcounts
#' @param cluster_id: ID used for finding clusters of cells
#' @export
export_ATAC_scANANSE <- function(seurat_object,
                                   min_cells = 50,
                                   output_dir = '~/',
                                   ATAC_peak_assay = "peaks",
                                   cluster_id = 'seurat_clusters'
) {
  dir.create(file.path(output_dir))
  Idents(seurat_object) <- cluster_id
  
  peak_count_lists = list()
  cluster_names = list()
  i = 1
  for (cluster in levels(Idents(seurat_object))){
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells){
      print(paste0('gather data from ', cluster, ' with ', n_cells, ' cells'))
      cluster_names[i] <- cluster
    
      #lets grab the ATAC signal
      mat <- seurat_object_cluster@assays[[ATAC_peak_assay]]@counts
      mat <- as.matrix(rowSums(as.matrix(mat)))
      colnames(mat) <- cluster
      peak_count_lists[i] <- as.data.frame(mat)
      i = i + 1
    }else{print(paste0(cluster,' less than ',min_cells,' skipping'))}
  }
  
  #ATAC peak matrix
  activity_matrix <- do.call(rbind, peak_count_lists)
  activity_matrix <- as.data.frame(activity_matrix)
  activity_matrix <- t(activity_matrix)
  colnames(activity_matrix) <- cluster_names
  #output the peaks as a bed file:
  peaks <- rownames(mat)
  peaks <- str_split_fixed(peaks, '-', 3)
  peaknames <- paste0(peaks[,1],":",peaks[,2],'-',peaks[,3])
  rownames(activity_matrix) <- peaknames
  activity_matrix <- as.data.frame(activity_matrix)
  activity_matrix$average <- round(rowMeans(activity_matrix))
  
  Peak_file <-paste(output_dir, "Peak_Counts.tsv", sep = '/')
  write.table(activity_matrix, Peak_file, sep = '\t', quote = F)
}

#' config_scANANSE
#'
#' This functions generates a sample file and config file for running Anansnake based on the seurat object
#' @param seurat object
#' @param min_cells: minimum of cells a cluster needs to be exported
#' @param output_dir: directory where the files are outputted
#' @param cluster_id: ID used for finding clusters of cells
#' @param additional_contrasts: additional contrasts to add between clusters within cluster_ID
#' @export
config_scANANSE <- function(seurat_object,
                                   min_cells = 50,
                                   output_dir = '~/',
                                   cluster_id = 'seurat_clusters',
                                   additional_contrasts = 'None'
) {
  dir.create(file.path(output_dir))
  Idents(seurat_object) <- cluster_id
  print('calculating CPM')
  seurat_object <- NormalizeData(seurat_object, 
                                 assay  = RNA_count_assay, 
                                 normalization.method = 'RC', 
                                 scale.factor = 1e6
  )
  
  Peak_file <-paste(output_dir, "Peak_Counts.tsv", sep = '/')
  count_file <- paste(output_dir, "RNA_Counts.tsv", sep = '/')
  CPM_file <-paste(output_dir, "TPM.tsv", sep = '/')
  cluster_names = list()
  i = 1
  for (cluster in levels(Idents(seurat_object))){
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells){
      cluster_names[i] <- cluster
    }}
  
  #lets generate the snakemake sample file
  sample_names <- c(cluster_names, 'average')
  sample_file_df <- as.data.frame(sample_names)
  sample_file_df <- as.data.frame(t(sample_file_df))
  colnames(sample_file_df) <- 'sample'
  sample_file_df$assembly <- 'hg38'
  sample_file_df$anansesnake <- sample_file_df$sample
  sample_file_location <- paste0(output_dir,"/samplefile.tsv")
  
  write.table(sample_file_df,sample_file_location,
              sep='\t',
              row.names = FALSE,
              quote = F)
  #lets generate the snakemake config file
  contrast_list <- as.list(paste0('anansesnake_',cluster_names,'_average'))
  if (type(additional_contrasts) == 'list'){print('adding additional contrasts')
    additional_contrasts <- paste0('anansesnake_',additional_contrasts)
    contrast_list = c(contrast_list,additional_contrasts)}
  
  file=paste0(output_dir,"/config.yaml")
  
  lines= c("genome: hg38 \n",
    "database: gimme.vertebrate.v5.0 \n",
    "jaccard: 0.1 \n",
    "edges: 500_000 \n",
    "padj: 0.05 \n",
    'plot_type: "png" \n',
    "get_orthologs: false \n",
    "contrasts: \n")
  
  string= paste(lines, collapse = '')
  cat(string,file=paste0(output_dir,"/config.yaml"),append = T)
  for (contr in contrast_list){
    cat(paste0('  - "',contr, '"'),"\n",file=paste0(output_dir,"/config.yaml"),append = T)
  }
  }

#' DEGS_scANANSE
#'
#' Calculate the differential genes needed for ananse influence
#' @param seurat object
#' @param min_cells: minimum of cells a cluster needs to be exported
#' @param output_dir: directory where the files are outputted
#' @param cluster_id: ID used for finding clusters of cells
#' @param additional_contrasts: additional contrasts to add between clusters within cluster_ID
#' @export
DEGS_scANANSE <- function(seurat_object,
                                      min_cells = 50,
                                      output_dir = '~/',
                                      cluster_id = 'seurat_clusters',
                                      additional_contrasts = 'None'
) {
  dir.create(file.path(output_dir))
  Idents(seurat_object) <- cluster_id
  cluster_names = list()
  i = 1
  for (cluster in levels(Idents(seurat_object))){
    seurat_object_cluster <- subset(x = seurat_object, idents = cluster)
    #only use ANANSE on clusters with more than 50 cells
    n_cells <- dim(seurat_object_cluster)[2]
    if (n_cells > min_cells){
      cluster_names[i] <- cluster
    }}
  
  #lets generate the snakemake config file
  contrast_list <- as.list(paste0('anansesnake_',cluster_names,'_average'))
  if (type(additional_contrasts) == 'list'){print('adding additional contrasts')
    additional_contrasts <- paste0('anansesnake_',additional_contrasts)
    contrast_list = c(contrast_list,additional_contrasts)}
   
  #lets calculate all the DEGs for the provided contrasts
  dir.create(file.path(paste0(output_dir,'/deseq2/')))
  
  for (contr in contrast_list){
    print(paste0('calculating DEGS for contrast ',contr))
    comparison1 <- str_split(contr, "_")[[1]][2]
    comparison2 <- str_split(contr, "_")[[1]][3]
    
    DEG_file <- paste0(output_dir,'/deseq2/','hg38-anansesnake_',comparison1,'_', comparison2,'.diffexp.tsv')
    
    if (file.exists(DEG_file)){print('skip')} else {
      
      if (comparison2 == 'average'){
        DEGS <- FindMarkers(seurat_object, 
                            ident.1 = comparison1,
                            logfc.threshold = 0.1,  
                            min.pct = 0.05, 
                            return.thresh = 0.1)
      }else{DEGS <- FindMarkers(seurat_object, 
                                ident.1 = comparison1, 
                                ident.2 = comparison2, 
                                logfc.threshold = 0.1,  
                                min.pct = 0.05,
                                return.thresh = 0.1)}
      DEGS <- DEGS[c('avg_log2FC','p_val_adj')]
      colnames(DEGS) <- c('log2FoldChange','padj')
      write.table(DEGS,DEG_file, sep = '\t', quote = F)
    }
  }
}

#' import_seurat_scANANSE
#'
#' import the influences from a anansnake directory into a seurat object
#' @param seurat object
#' @param anansnake_inf_dir influence directory generated by anansnake
#' @export
import_seurat_scANANSE <- function(seurat_object,
                                   cluster_id = 'seurat_clusters',
                                   anansnake_inf_dir = 'None'
){
  
  files <- list.files(path=anansnake_inf_dir, pattern="*.tsv", full.names=TRUE, recursive=FALSE)
  GRN_files <- list.files(path=anansnake_inf_dir, pattern="_diffnetwork.tsv", full.names=TRUE, recursive=FALSE)
  files <- files[!files %in% GRN_files]
  
  influence_scores_avg = list()
  influence_targets = list()
  i = 1
  
  for (file in files){
    cell_target <- stringr::str_split(basename(file),"_")[[1]][2]
    cell_source <- stringr::str_split(basename(file),"_")[[1]][3]
    cell_source <- stringr::str_replace(cell_source,'.tsv','')
    in_df <- read.table(file, header = T)
    in_df <- in_df[c('factor','influence_score')]
    colnames(in_df) <- c('factor',cell_target)
    if (cell_source == 'average'){
      influence_scores_avg[[i]] <- as.data.frame(in_df)
      influence_targets[[i]] = cell_target
      i = i + 1
    }
  }
  
  avg_df <- influence_scores_avg %>% purrr::reduce(dplyr::full_join, by = "factor", copy = T)
  avg_df[is.na(avg_df)] = 0
  rownames(avg_df) <- avg_df$factor
  avg_df$factor <- NULL
  
  #add the TF intensities to the seurat object
  TF_df <- t(avg_df)
  cell_signal_list <- list()
  i = 1
  #get cell IDs from the pbmcs
  for (cell_type in rownames(TF_df)){
    relevant_IDS <- seurat_object[,seurat_object[[cluster_id]] ==cell_type][[cluster_id]]
    TF_signal <- TF_df[rownames(TF_df) == cell_type,]
    TF_signal <- as.data.frame(TF_signal)
    TF_signal <- t(TF_signal)
    TF_signal <- TF_signal[rep(seq_len(nrow(TF_signal)), each = nrow(relevant_IDS)), ]
    rownames(TF_signal) <- rownames(as.data.frame(relevant_IDS))
    cell_signal_list[[i]] <- TF_signal
    i = i + 1
  }
  TF_array <- do.call("rbind", cell_signal_list)
  
  #add cell IDs that do not have ANANSE signal and give them a NA vallue
  cell_barcodes_all <- rownames(as.data.frame(Idents(object = seurat_object)))
  cell_barcodes_missing <- cell_barcodes_all[!cell_barcodes_all %in% rownames(TF_array)]
  cell_barcodes_missing <- as.data.frame(cell_barcodes_missing)
  for (TF in colnames(TF_signal)){
    cell_barcodes_missing[[TF]] <- NA
  }
  rownames(cell_barcodes_missing) <- cell_barcodes_missing$cell_barcodes_missing
  cell_barcodes_missing$cell_barcodes_missing <- NULL
  
  TF_output <- t(rbind(TF_array,cell_barcodes_missing))
  
  seurat_object[['ANANSE']] <- TF_output %>% Seurat::CreateAssayObject(.)
  return(list(seurat_object,avg_df))
}